packages/components/CONTRIBUTING.md

```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
 
 The following is a set of guidelines for contributing to the `@wordpress/components` package to be considered in addition to the general ones described in our [Contributing Policy](/CONTRIBUTING.md).
 
-## Examples
+## Core principles
 
-Each component needs to include an example in its README.md file to demonstrate the usage of the component.
+Contributions to the `@wordpress/components` package should follow a set of core principles and technical requirements.
 
-These examples can be consumed automatically from other projects in order to visualize them in their documentation. To ensure these examples are extractable, compilable and renderable, they should be structured in the following way:
+This set of guidelines should apply especially to newly introduced components. It is, in fact, possible that some of the older component don't respect some of these guidelines for legacy/compat reasons.
 
--   It has to be included in a `jsx` code block.
--   It has to work out-of-the-box. No additional code should be needed to have working the example.
--   It has to define a React component called `My<ComponentName>` which renders the example (i.e.: `MyButton`). Examples for the Higher Order Components should define a `MyComponent<ComponentName>` component (i.e.: `MyComponentWithNotices`).
+### Compatibility
+
+The `@wordpress/components` package includes components that are relied upon by many developers across different projects. It is, therefore, very important to avoid introducing breaking changes.
+
+In these situations, one possible approach is to "soft-deprecate" a given legacy API. This is achieved by:
+
+1. Removing traces of the API from the docs, while still supporing it in code;
+2. Updating all places in Gutenberg that use that API;
+3. Adding deprecation warnings (only after the previous point is completed, otherwide the Browser Console will be polluted by all those warnings and some e2e tests may fail).
+
```

Comment:
I'd suggest adding a note about prefixing new components and props with `__unstable` or `__experimental`. I'm not sure if we all agree with that, but that's something I often suggest in code reviews.

Also, linking to the specific docs would be useful.

```suggestion

When adding new components or new props to existing components, it's recommended to prefix them with `__unstable` or `__experimental` until they're stable enough to be exposed as part of the public API.

Learn more on [How to preserve backward compatibility for a React Component](https://github.com/WordPress/gutenberg/blob/trunk/docs/how-to-guides/backward-compatibility/README.md#how-to-preserve-backward-compatibility-for-a-react-component) and [Experimental and Unstable APIs](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md#experimental-and-unstable-apis)

```

Reply:
That's a good default to go with. In the context of components, it's going to be experimental in most of the cases ❤️

Reply:
Makes a lot of sense, thank you @diegohaz for suggesting this approach!

@gziolo, out of curiosity — when adding links to other files in the repo, I omitted the `(https://github.com/WordPress/gutenberg/blob/trunk/` prefix.

Do you have a preference around the link format?

Reply:
At some point @sarayourfriend updated all links to start with: `https://github.com/WordPress/gutenberg/blob/HEAD/`. It might be related to the switch from `master` to `trunk` though.

Some links get special treatment in the Block Editor Handbook, see more in https://github.com/WordPress/gutenberg/tree/trunk/docs/contributors/documentation#using-links.

Reply:
Thank you both for the details explanation!

Following the documentation linked by @gziolo , I've decided to update all internal link to the "absolute path" format in [`8c494c8` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/8c494c83282cb1a9cf68f0bb51597d91d5f1abb5)